Document query builder v2 (#100)
0xPolygonID/id-docs: docs/verifier/query-builder.md

id: query-builder
title: Query Builder
sidebar_label: Query Builder
description: Learn how to use the Query Builder.
keywords:
  - docs
  - polygon id
  - ID holder
  - issuer
  - verifier
  - query builder
  - credentialAtomicQuerySigV2
  - credentialAtomicQuerySigV2OnChain
  - credentialAtomicQueryMTPV2
  - credentialAtomicQueryMTPV2OnChain
  - credentialAtomicQueryV3
  - credentialAtomicQueryV3OnChain
  - selective disclosure
  - proof of uniqueness
  - nullifier
---

import useBaseUrl from '@docusaurus/useBaseUrl';

The [Query Builder](https://schema-builder.polygonid.me/query-builder) is a dynamic, user-friendly tool designed for crafting verification queries. It simplifies the process by providing a comprehensive form, encompassing all necessary fields for creating detailed queries.

This tool enables verifiers to establish specific rules for identity verification, catering to a range of criteria from organizational membership to age requirements. Polygon ID enhances this process, offering a secure and streamlined way for users to authenticate their identity based on these criteria.

> With its standardized interface, the Query Builder ensures uniformity, consistency, and minimizes errors in query creation across various users and organizations. It also presents a more intuitive alternative to the [ZK Query Language](./verification-library/zk-query-language.md), making query creation accessible even for those without in-depth technical knowledge.

Reviewer: > It also presents a more intuitive alternative to the ZK Query Language.

It's not an alternative to the ZK Query Language, it's a tool to create queries using the ZK Query Language without having to learn its specifics.

Author: What about:

> It also presents a more intuitive alternative to manually writing queries for the ZK Query Language.

This was the implicit meaning already, but you're right.

Author: Done

<div align="center">
<img src={useBaseUrl("/img/query-builder.png")} align="center" width="600"/>
</div>
<br/>

## Building a query

### URL to JSON-LD Context

To create a query, it is essential to reference the correct schema, specifically its JSON-LD Context's URL. Users have two options: they can either enter a custom URL directly or utilize the Schema Explorer. By beginning to type the name (or any other keyword) of the desired schema in the Schema Explorer, a list of available schemas will appear, allowing the user to select the appropriate one.

<div align="center">
<img src={useBaseUrl("/img/query-builder-url.png")} align="center" width="600"/>
</div>
<br/>

### Schema type

> The JSON-LD Context derived from a schema in the Schema Explorer is limited to a single type. However, a custom JSON-LD Context may encompass multiple types. This particular field enables users to select the appropriate type for constructing their query. As a user-friendly feature, in cases where the JSON-LD Context is single-typed, the field is pre-populated automatically.

Reviewer: I would remove the `derived` keyword, the JSON-LD Context is not derived from anything...

> The JSON-LD Context from a schema in the Schema Explorer is limited to a single type.

Also, here `pre-populated` looks weird. When you fill a form you are filling the inputs, but not populating them:

> As a user-friendly feature, in cases where the JSON-LD Context is single-typed, the field is *pre-filled* automatically.

Author: I don't agree with these statements.

> I would remove the `derived` keyword, the JSON-LD Context is not derived from anything...

The JSON-LD Context is indeed derived from the schema in the Schema Explorer. The source of truth is the schema, the JSON LD Context is just a serialization of the schema. I agree though that this is an implementation detail that users don't care about.

> Also, here `pre-populated` looks weird. When you fill a form you are filling the inputs, but not populating them

[Populate](https://www.google.com/search?q=%22pre-populate+a+form%22) is perfectly correct in this context.

In any case, I added the two changes because I like you.

### Attribute field

Upon specifying the URL and type, the interface displays a hierarchical tree structure that showcases the attributes associated with the selected type. Queries are based on the values of these attributes, which are stored within a user's wallet as part of their credentials.

Credentials consist of various primitive attributes such as strings, integers, booleans, etc. These primitive attributes may be structured in a nested format using object attributes (represented as folders in the tree). Since queries are only executable against primitive attributes, object attributes (folders), are not selectable within the tree.

<div align="center">
<img src={useBaseUrl("/img/query-builder-attribute-field.png")} align="center" width="600"/>
</div>
<br/>

### Proof type

> This field allows specifying the verification method used by the issuer at the time of issuing the credential, which is essential for successful query verification. While creating a query, users must carefully choose a Proof Type that aligns with the original method selected by the issuer when issuing the credential. The two options available are:

Reviewer: > which is essential for *a* successful query verification

Author: Done

> - **Signature-based (SIG):** Validates claims based on Baby JubJub (BJJ) Key Signatures. This method is straightforward, relying on the issuer's signature for authentication. It's fast and doesn't involve on-chain transactions.

Reviewer: > *Verifies* claims based on Baby JubJub (BJJ) key signatures

Author: Done

> - **Merkle Tree Proof (MTP):** Verifies claims through their inclusion in a Merkle Tree, part of the issuer’s on-chain identity. This approach not only bolsters security but also provides a timestamp, confirming the claim's existence at a specific historical moment.

Reviewer: > Verifies claims through their inclusion in a Merkle Tree, *which is* part of the issuer’s on-chain identity.

Author: Done

> - **Merkle Tree Proof (MTP):** Verifies claims through their inclusion in a Merkle Tree, which is part of the issuer’s on-chain identity. This approach not only bolsters security but also provides a timestamp, confirming the claim's existence at a specific historical moment.

Reviewer 2: In general, for v3 circuit it is possible that verifier doesn't know proof type which will be used. So `proofType` field will be missed in circuit and user will use the cred with any proof.

Author: At present, you have the option to select any proof type associated with V3 circuits, but not both or none (since it's radio button).

If I understand correctly, you are suggesting the addition of a third radio button option labeled "Any." This would mean that when "Any" is selected, the Circuit ID input below will exclusively display V3 circuits. Is my understanding correct?

Reviewer 2: This would mean that when "Any" is selected, the Circuit ID input below will exclusively display V3 circuits. Is my understanding correct?  - yes.

Author: c.c @elias-garcia @Figmintojas

Reviewer: I've created another task in Jira for this.

### Circuit ID

The Query Builder incorporates 6 [Circom](https://docs.circom.io/) circuits, each designed for specific verification scenarios. Selection of the appropriate Circuit ID depends on the verification context (off-chain or on-chain) and the proof type used when the claim was issued. Below is a guide to these circuits:

1. **credentialAtomicQuerySigV2:** Verifies off-chain if a claim, signed by the Issuer, meets the verifier's criteria.
2. **credentialAtomicQuerySigV2OnChain:** Similar to SigV2, but for on-chain verification (smart contract verifiers).
3. **credentialAtomicQueryMTPV2:** For off-chain verification of claims added to the issuer's claims Merkle Tree.

> 4. **credentialAtomicQueryMTPV2OnChain:** Applies MTPV2 logic for on-chain scenarios (smart contract verifiers).

Reviewer: > 4. **credentialAtomicQueryMTPV2OnChain:** Applies MTP V2 logic for on-chain scenarios (smart contract verifiers).

Author: Done

5. **credentialAtomicQueryV3:** A versatile circuit for claims either signed by the Issuer or included in the Issuer's state, suitable for off-chain verification.
6. **credentialAtomicQueryV3OnChain:** The on-chain counterpart of V3, also accommodating both SIG and MTP proof types.

#### Circuit Selection Based on Proof Type

The available circuits in the Query Builder are filtered based on the selected Proof Type:

- **Signature-based (SIG) Proof Type:** For SIG, the list includes V2 SIG circuits (1 and 2) and the V3 circuits, as they support both SIG and MTP proof types.
- **Merkle Tree Proof (MTP) Proof Type:** With MTP, the options presented are V2 MTP circuits (3 and 4) and also the V3 circuits, compatible with both SIG and MTP.

It's important to consider the proof type as it directly influences the circuit options provided, ensuring alignment with the verification method (off-chain or on-chain via a smart contract) and the specific requirements of the query.

:::caution
V3 circuits are experimental as of this writing, so users should exercise caution and stay informed about their development and stability.
:::

### Proof of Uniqueness

> The Proof of Uniqueness enhances security and integrity in contexts requiring [sybil resistance](https://en.wikipedia.org/wiki/Sybil_attack). Using this feature is optional when building a query and it's exclusively supported by v3 circuits. It's managed through two fields in the Query Builder form:

Reviewer: > This feature is optional and it's exclusively supported by v3 circuits.

Author: Done

> 1. **Check Proof of Uniqueness:** This checkbox is initially unchecked. When selected, it enables the Proof of Uniqueness functionality.

Reviewer: >1. **Enable Proof of Uniqueness:** This checkbox is initially unchecked. When selected, it enables the Proof of Uniqueness *feature*.

Author: Done

> 2. **Nullifier Session ID:** A numeric input field that remains disabled unless the Proof of Uniqueness checkbox is checked. It requires the user to provide a nullifier session ID.

Reviewer: >2. **Nullifier Session ID:** A numeric input field that remains disabled unless the *enable Proof of Uniqueness checkbox* is checked. It requires the user to provide a nullifier session ID.

Author: This is the UI:

> A numeric input field that remains disabled unless the Proof of Uniqueness checkbox is checked. It requires the user to provide a nullifier session ID

> The nullifier session ID field is a critical component for calculating the SybilID, which is essential for establishing Proof of Uniqueness in a query. This ID serves as a verifier-specific session identifier, which might represent, for example, a voting ID. It is used in the SybilID calculation formula:

Reviewer 2: let's rename SybilID to NullfierID (we don't use term sybil publicly I guess)

Author: Done

`SybilID = hash(genesisID, credProfileNonce, schemaHash, verifierID, nullifierSessionID)`

> Providing the correct nullifier session ID ensures the query's integrity and uniqueness, conforming to the protocol's sybil resistant features:

Reviewer 2: correct nullifier session ID => it is just a random big integer. I think `correct` can mislead user.

Author: Done

> - **One Identity Per Session:** Each identity can obtain only one identifier per credential and verifier session ID (CRS), consistent across all Profiles and Genesis ID.

Reviewer 2: verifier session ID (CRS) => nullfier session id
```suggestion
- **One Credential Per Session:** Each identity can obtain only one Nullfier per specific credential, nullfier session id, consistent across all Profiles and Genesis ID that are used for the same verifier.
```

- **Brute-force Resistance:** Prevents brute-forcing to reveal Genesis ID or other values.

> - **Tracking Resistance:** Ensures Profile IDs and Genesis ID cannot be linked based on the Sybil ID.

Reviewer 2: ```suggestion
- **Tracking Resistance:** Ensures Profile IDs and Genesis ID cannot be linked based on the Nullifer ID.
```

- **Flexible Session Duration:** Allows the verifier to keep the session active as needed, without time-specific constraints.
- **Protection Against Collusion:** Prevents matching different identifiers as belonging to the same identity, even if the issuer and verifier collude.

### Query Type

The protocol allows selection between two query types:

1. **Condition:** This type allows the prover to demonstrate that a specific condition is met without revealing the underlying data. It maintains data privacy by cryptographically verifying the condition. An example is proving legal age for entry into a venue without disclosing the actual age.

> 2. **Selective disclosure:** Use this when the prover needs to reveal specific data stored in their credential. This option discloses the actual value of an attribute to the verifier. For instance, instead of just proving legal age, it would show the verifier the prover's actual age.

Reviewer: > 2. **Selective disclosure:** Use this when you need the prover to reveal specific data stored in user credential. This option discloses the actual value of an attribute to the verifier. For instance, instead of just requesting a proof proving legal age, it would disclose user's actual age to the verifier.

Author: The `prover` and the `user` are the same here. I'd not use two terms.

> 2. **Selective disclosure:** Use this when you need the prover to reveal specific data stored in their credential. This option discloses the actual value of an attribute to the verifier. For instance, instead of just requesting proof proving legal age, it would disclose the prover's actual age to the verifier.

Author: Done

Choose "Condition" for privacy-preserving verification, or "Selective disclosure" for cases where revealing specific information is necessary.

### Operator

In conditional queries, the "Operator" input defines the logical operation to be applied. The supported operators are:

- **Equal:** Checks if the value is equal to the specified value.
- **Not Equal:** Verifies if the value is not equal to the specified value.

> - **Less Than:** Assesses if the value is less than the given value.

Reviewer: > - **Less Than:** Assesses *that* the value is less than the given value.

Or even maybe:

> - **Less Than:** Assesses the value is less than the given value.

Author: I prefer the first. Done.

- **Greater Than:** Evaluates if the value is greater than the specified value.
- **Matches one of the values:** Determines if the value is within a given set of values.
- **Matches none of the values:** Checks if the value is not within a specified set.

> Choose the appropriate operator based on the condition you need to validate.

Reviewer: > Choose the appropriate operator based on the condition you *want* to validate.

Author: Done

### Attribute value

> The "Attribute value" field allows users to input the specific value or range of values to be compared with the selected attribute in the prover's credential. This field adapts to the type of the chosen attribute, accommodating various data types such as strings, numbers, booleans, dates, and more. This ensures that the value provided is compatible with the selected operator and the nature of the attribute.

Reviewer: > The "Attribute value" field allows users to input the specific value or range of values *used to* compare with the selected attribute in the *user's* credential.

The prover just takes the required data from the credential and the query as an input and it outputs the proof. The credentials always belong to the users.

Author: > The prover just takes the required data from the credential and the query as an input and it outputs the proof. The credentials always belong to the users.

Man, what a mess! In your sentence above the `prover` and the `user` are the same, right? I mean the prover is not software, it's the person that (using software) proves something.

Also, I think we have to avoid using the word "user". It's ambiguous because we use it both for the verifier and the prover (and maybe even the issuer...).

I'd rephrase to:

> The "Attribute value" field allows verifiers to input the specific value or range of values used to compare with the selected attribute in the prover's credential.

Author: Done

### Issuer DID

This field in the query form is used to specify the Decentralized Identifier (DID) of the issuer that is trusted by the verifier. By entering a specific DID, the verifier can confirm that the credential presented by the prover was indeed issued by the specified issuer.

:::caution

Using `*` as a wildcard to accept credentials from any issuer is possible but not recommended, particularly in production environments. This approach may compromise the query's security and integrity, as it will validate credentials issued by any issuer.

:::

:::caution

> Currently, the system supports specifying only a single DID. This means users have the choice to either provide one specific issuer DID or use `*` to allow any issuer. However, the former is strongly advised for maintaining robust verification standards.

Reviewer 2: not the system, but the query builder. On protocol level you can have more then one specified issuer

Author: I thought this was a limitation in the protocol. I don't understand why we have this limited in the Query Builder then...

c.c @elias-garcia @javip97

Author: Done

Reviewer: We limited it in the query builder due to time constraints, but I'm aware that the protocol supports more than one issuer. I've created a ticket for this in Jira.

:::

### Skip Revocation Check

This checkbox, when enabled, instructs the circuit to bypass the verification step that checks whether a credential has been revoked.

:::caution

> Opting to skip the revocation check is generally not advised. This check is a crucial security feature, ensuring that the credential being used is still valid and has not been rescinded by the issuer. Skipping this step can lead to the acceptance of outdated or invalidated credentials, potentially compromising the integrity of the verification process and the trustworthiness of the results.

Reviewer: > This check is a crucial security feature, ensuring that the credential being used is still valid and has not been *revoked* by the issuer.

Author: Done

:::

## Verifying a query

> 

Reviewer: > After setting up the query in the Query Builder, the verification can be performed (which differs for off-chain and on-chain queries).

Maybe it's worth adding that the verification screen you'll see depends on the `Circuit ID` field of the query builder form?

> After setting up the query in the Query Builder, verification can be performed, which differs for off-chain and on-chain queries.

Reviewer: ```suggestion
After setting up the query in the Query Builder, verification can be performed, which differs according to the selected circuit (off-chain or on-chain).
```
